DiscoPoP/DiscoPoP.cpp

```diff
@@ -0,0 +1,2431 @@
+
+        // // Helper functions
+        // bool isaCallOrInvoke(Instruction *BI);
+        // bool sanityCheck(BasicBlock *BB);
+        // void collectDebugInfo();
+        // void processStructTypes(string const &fullStructName, MDNode *structNode);
+        // DIGlobalVariable *findDbgGlobalDeclare(GlobalVariable *V);
+        // Value *getOrInsertVarName(string varName, IRBuilder<> &builder);
+        // Value *findStructMemberName(MDNode *structNode, unsigned idx, IRBuilder<> &builder);
+        // Type *pointsToStruct(PointerType *PTy);
+        // Value *determineVariableName(Instruction *const I);
+
+        // // Control flow analysis
+        // void CFA(Function &F, LoopInfo &LI);
+
+        // // Callback Inserters
+        // //void insertDpInit(const vector<Value*> &args, Instruction *before);
+        // //void insertDpFinalize(Instruction *before);
+        // void instrumentStore(StoreInst *toInstrument);
+        // void instrumentLoad(LoadInst *toInstrument);
+        // void insertDpFinalize(Instruction *before);
+        // void instrumentFuncEntry(Function &F);
+        // void instrumentLoopEntry(BasicBlock *bb, int32_t id);
+        // void instrumentLoopExit(BasicBlock *bb, int32_t id);
+
+        // int64_t uniqueNum;
+
+        // // Callbacks to run-time library
+        // Function *DpInit, *DpFinalize;
+        // Function *DpRead, *DpWrite;
+        // Function *DpCallOrInvoke;
+        // Function *DpFuncEntry, *DpFuncExit;
+        // Function *DpLoopEntry, *DpLoopExit;
+
+        // // Basic types
+        // Type *Void;
+        // IntegerType *Int32, *Int64;
+        // PointerType *CharPtr;
+
+        // // Control flow analysis
+        // int32_t loopID;
+        // int32_t fileID;
+
+        // // Output streams
+        // ofstream ocfg;
+
+        // // Export Module M from runOnModule() to the whole structure space
+        // Module *ThisModule;
+        // LLVMContext *ThisModuleContext;
+
+        // map<string, Value *> VarNames;
+        // set<DIGlobalVariable *> GlobalVars;
+        // map<string, MDNode *> Structs;
+    // };
+// }  // namespace
+
+// char DiscoPoP::ID = 0;
+
+// static RegisterPass<DiscoPoP> X("DiscoPoP", "DiscoPoP: finding potential parallelism.", false, false);
+
+// static void loadPass(const PassManagerBuilder &Builder, legacy::PassManagerBase &PM)
+// {
+//     PM.add(new LoopInfoWrapperPass());
+//     PM.add(new DiscoPoP());
+// }
+
+// static RegisterStandardPasses DiscoPoPLoader_Ox(PassManagerBuilder::EP_OptimizerLast, loadPass);
+// static RegisterStandardPasses DiscoPoPLoader_O0(PassManagerBuilder::EP_EnabledOnOptLevel0, loadPass);
+
+// FunctionPass *createDiscoPoPPass()
+// {
+//     if (DP_DEBUG)
+//     {
+//         errs() << "create DiscoPoP \n";
+//     }
+//     initializeLoopInfoWrapperPassPass(*PassRegistry::getPassRegistry());
+//     return new DiscoPoP();
+// }
+
+StringRef DiscoPoP::getPassName() const
+{
+    return "DiscoPoP";
+}
+
```

Comment:
While it is trivial and could be moved to the header for this reason, it nevertheless is an implementation of the declared function. I would suggest to keep it as is so that `DiscoPop.cpp` contains all implementations, while the header only declares functions.

Reply:
ok, I can get behind that. My thought was that it is a) trivial, and b) the name of the pass is very much like a part of the interface and it feels right to put it to the header.

But I agree. From an implementation view it is good practice to keep declaration and definition separated.

Reply:
That's a good point, though... 
Feel free to decide how we proceed, i think both options make sense.